tests/RegistrationTest.php

```diff
@@ -0,0 +1,160 @@
+                    'accept-tos' => true,
+                ])
+            );
+        }
+    }
+
+    public function testNoregLocal()
+    {
+        $local = NoregLocal::first()->local;
+
+        $this->assertRedirectedToRoute('user.create', [], [],
+            $this->call('POST', route('user.store'), [
+                'local' => $local,
+                'domain' => $this->domain,
+                'password' => 'test',
+                'password-verify' => 'test',
+                'accept-tos' => true,
+            ])
+        );
+    }
+
+    public function testPrivateDomain()
+    {
+        $domain = Domain::findByNameOrFail('private.test.test')->uuid;
+
+        $this->assertRedirectedToRoute('user.create', [], [],
+            $this->call('POST', route('user.store'), [
+                'local' => 'this.surely.wont.exist.in.the.tests',
+                'domain' => $domain,
+                'password' => 'test',
+                'password-verify' => 'test',
+                'accept-tos' => true,
+            ])
+        );
+    }
+
+    public function testShowFormAsGuest()
+    {
+        $this->assertResponseOk($this->call('GET', route('user.create')));
+    }
+
+    public function testShowFormAsUser()
+    {
+        $this->assertRedirectedToRoute('dashboard', [], [],
+            $this
+                ->actingAs(User::findByAddressOrFail('user@test.test'))
+                ->call('GET', route('user.create'))
+        );
+    }
+
+    /**
+     * Tests that fail with PHP's filter_var have been disabled.
+     */
+    public function testSuccess()
+    {
+        $locals = [
+            'prettyandsimple',
+            'very.common',
+            'other.email-with-dash',
+            'x',
+            // '"much.more unusual"',
+            '"very.unusual.@.unusual.com"',
+            '"very.(),:;<>[]\".VERY.\"very@\\ \"very\".unusual"',
+            'example-indeed',
+            '#!$%&\'*+-/=?^_`{}|~',
+            // '"()<>[]:,;@\\\"!#$%&\'-/=?^_`{}| ~.a"',
+            // '" "',
+        ];
+
+        foreach ($locals as $local) {
+            $this->assertRedirectedToRoute('login', [], [],
+                $this->call('POST', route('user.store'), [
+                    'local' => $local,
+                    'domain' => $this->domain,
+                    'password' => 'test',
+                    'password-verify' => 'test',
+                    'accept-tos' => true,
+                ])
+            );
+        }
+    }
+
+    public function testUncheckedTos()
+    {
+        $this->assertRedirectedToRoute('user.create', [], [],
+            $this->call('POST', route('user.store'), [
+                'local' => 'this.surely.wont.exist.in.the.tests',
+                'domain' => $this->domain,
+                'password' => 'test',
+                'password-verify' => 'test',
+            ])
+        );
+    }
+
+    public function testUnknownDomain()
+    {
+        $this->assertRedirectedToRoute('user.create', [], [],
```

Comment:
Would it not be a good idea to check what the exception here is too alongside the redirect?